### PR TITLE
Provide the ability to select the series when adding a machine.

### PR DESCRIPTION
**Pull request by frankban**
Also fix how we show the constraints, so that only constraints applicable to the current provider type are available to be set. And when adding containers, constraints and action button are not displayed until the user selects the container type.

This branch also includes:
- fixing a bug for which constraints are shown in the inspector scale-up view were always disabled;
- fixing a UX artifact in charm/bundle detail headers when the corresponding entity is not assigned to any channel;
- fixing console.log not working in tests.

Fixes https://github.com/juju/juju-gui/issues/1788

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6969/

**Comment by frankban** [comment]
QA: use the machine view to create machines and containers, with different series and constraints. Use different providers (ec2, azure, lxd) and check that the proper constraints are displayed for each cloud.
Scale up a unit and contextually ad constraints.

File: jujugui/static/gui/src/app/components/constraints/constraints.js [hunk 1]
```diff
@@ -48,21 +49,76 @@ YUI.add('constraints', function() {
...
-          <label htmlFor="cpu-constraint"
-            className="constraints__label">
+      const props = this.props;
+      const disabled = props.disabled;
+      let series;
+      // Only allow selecting a series if there is no unit already assigned.
+      // If there is a unit, the machine must have the same series as the unit.
+      if (!props.hasUnit) {
+        // Generate a list of series options.
+        const machineSeries = window.jujulib.SERIES.filter(ser => {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/constraints/constraints.js]: It's a little weird that we have "bundle" in the series list, I wonder if we should provide a const without that in there instead...

File: jujugui/static/gui/src/app/components/constraints/constraints.js [hunk 2]
```diff
@@ -104,7 +172,64 @@ YUI.add('constraints', function() {
...
+          case 'lxc':
+            parts = [];
+            break;
+          case 'lxd':
+            parts = [];
+            break;
+        }
+      } else {
+        // This is a top level machine, constraints are supported based on the
+        // current provider type.
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/constraints/constraints.js]: Thanks for adding this

File: karma-mocha-old.conf.js.tmpl [hunk 3]
```diff
@@ -11,6 +11,10 @@ module.exports = function(config) {
     },
 
 
+    // Set up the log level so that console.log messages are visible.
+    browserConsoleLogOptions: {level: 'log'},
+
+
```
Comment (hatched) [inline karma-mocha-old.conf.js.tmpl]: You left the extra lines in this one ;)

**Comment by kadams54** [comment]
QA notes:

When scaling up, if I select "New machine" in the "Move to" menu, I see the constraints; however, if I change that to an existing machine, the constraints are still displayed. Also: I'm guessing we don't ask about the series when creating a new machine in the scale up UI because that's constrained by the service that's being deployed to the machine, correct? If that's the case than it seems like we should offer the choice when dealing with a multi-series charm (constrained, of course, to the series on the charm).

File: jujugui/static/gui/src/app/components/constraints/_constraints.scss [hunk 4]
```diff
@@ -11,4 +11,9 @@
     margin: 5px 10px 10px;
     width: calc(100% - 20px);
   }
+
+  &__select {
+    margin: 5px 10px 10px 10px !important;
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/constraints/_constraints.scss]: I know we can do better. I am open to suggestions.

File: jujugui/static/gui/src/app/components/constraints/constraints.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/constraints/constraints.js]: From the URL perspective bundle is a series. I agree we could provide the list of series.

File: karma-mocha-old.conf.js.tmpl (hunk 3 above)
Comment (frankban) [inline karma-mocha-old.conf.js.tmpl]: Let me remove them ;-)

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6982/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6983/

File: jujugui/static/gui/src/app/components/constraints/constraints.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/components/constraints/constraints.js]: Ok fixed.

File: jujugui/static/gui/src/app/components/constraints/constraints.js [hunk 5]
```diff
@@ -48,21 +49,76 @@ YUI.add('constraints', function() {
...
-            className="constraints__label">
+      const props = this.props;
+      const disabled = props.disabled;
+      let series;
+      // Only allow selecting a series if there is no unit already assigned.
+      // If there is a unit, the machine must have the same series as the unit.
+      if (!props.hasUnit) {
+        // Generate a list of series options.
+        const machineSeries = window.jujulib.SERIES.filter(ser => {
+          return ser !== window.jujulib.BUNDLE_SERIES;
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/constraints/constraints.js]: Done.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6984/

**Comment by frankban** [comment]
@kadams54 I think you've found a bug that has been there forever! Fixed it. For the series in the charm, I am not sure that, once a charm unit is deployed, you can still select which series to use to create additional units. This can be an investigation and an enhancement for later. For now, it's just a corner case and I'd not focus on that.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6985/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6989/

File: jujugui/static/gui/src/app/jujulib/urls.js [hunk 6]
```diff
@@ -352,5 +352,9 @@ var module = module;
   const jujulib = exports.jujulib;
   jujulib.URL = URL;
   jujulib.SERIES = SERIES;
+  jujulib.BUNDLE_SERIES = BUNDLE_SERIES;
+  jujulib.CHARM_SERIES = SERIES.filter(series => {
```
Comment (hatched) [inline jujugui/static/gui/src/app/jujulib/urls.js]: Thanks!

**Comment by kadams54** [comment]
Re-QA'd and reviewed. I took a stab at fixing the constraint-related CSS issues; I found there were quite a few tricky situations where I ended up having to construct very specific selectors to avoid !important. I proposed those CSS fixes as a PR on @frankban's branch, here:

https://github.com/frankban/juju-gui/pull/1

@frankban if they look good to you, I think you should be able to merge my PR and then have those changes automatically pulled into this PR. I've never tried this though? Very curious to see if it'll work...

Aside from the CSS issues, there are still weird things happening with constraints in the scale up UI:

If you select an existing container, constraints show up (and I don't think you should be able to set constraints on something that's already been created):

selection_011

If you select a new container, constraints show up before selecting a container type:

selection_010

And if you select LXD for your container type, there are no constraints but the "Define constraints" header is still there:

selection_012

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7020/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7021/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7022/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7028/

**Comment by kadams54** [comment]
QA is OK, much better, thanks!

**Comment by frankban** [comment]
Thanks for the reviews and for QAing this.
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Build failed: Attempt to land pull request failed
        build url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge/2217

**Comment by frankban** [comment]
Seems a chromium disconnection? Retrying...
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**